### ISSUE TITLE
[Documentation/entity-facet.md]  PredefinedType as USERDEFINED and custom word in ObjectType 

### ISSUE THREAD

**Issue by Martin15135215** [buildingSMART/IDS#178]
https://github.com/buildingSMART/IDS/blob/50f2ceda8ad7df6df312ffdb940e6a72be719552/Documentation/entity-facet.md?plain=1#L13

https://github.com/buildingSMART/IDS/blob/50f2ceda8ad7df6df312ffdb940e6a72be719552/Documentation/entity-facet.md?plain=1#L22

I thought that if you have your own categorisation, you write PredefinedType as USERDEFINED and in ObjectType you put your own designation. The sentences above suggest otherwise.

**Comment by Moult** [comment]
This is correct. However from the perspective of IDS, it "auto-fetches" the ObjectType if PredefinedType = USERDEFINED.

**Comment by Martin15135215** [comment]
I see :)

It just confused me when I read the Documentation.

**Comment by CBenghi** [comment]
There is probably the need to state in the documentation that 
`IDS auto-fetches the ObjectType if PredefinedType = USERDEFINED.`

**Comment by Moult** [comment]
It's a bit unfortunate to write this because this is such an important part of IFC. The only reason we have "USERDEFINED" is because that's what STEP / EXPRESS forces you to do.

**Comment by RickBrice** [comment]
@atomczak @Moult I'm struggling with this issue right now. Can you explain the difference between these two IDS specifications?

Spec 1.
![Image](https://github.com/user-attachments/assets/3bc4b413-135a-4b06-871a-f2802f0cecb1)

Spec 2.
![Image](https://github.com/user-attachments/assets/ec5f616c-87c4-4cfa-be96-e142f8ac08e8)

Given this entry in my IFC file

> #872=IFCBEAM('1L5c$BTMnCxhBJ7IXhvkC8',#14666,'Segment 1',$,'GFRP',#851,#871,$,.USERDEFINED.);

When checking my file against the IDS, the first specification signals an error, while the second specification indicates the requirement is satisfied.

What is the difference?

**Comment by Moult** [comment]
The fact that predefined type has to be set to USERDEFINED, instead of directly set to GFRP is a technical restriction of the schema - it has no semantics. From the point of view of IFC / IDS, predefined type = GFRP.

As a result, in IDS, you just say "give me a predefined type of GFRP", you cannot say "give me a USERDEFINED, then also GFRP". Don't forget, in IFC, the predefined type isn't just an attribute value, it can be inherited, so it could be set in either the relating type's predefined type, ElementType attribute, occurrence's predefined type, or ObjectType attribute (or ProcessType attribute). Again, these are all just under-the-hood details: the semantic meaning is just "it is GFRP", regardless of how it got there, via inheritance or otherwise.

So the second spec is correct.

If you _really_ want to specify it at the under-the-hood level for some reason, use the attribute facet.

**Comment by RickBrice** [comment]
@Moult Just for educational purposes, I tried to use the attribute facet and specify it at the under-the-hood level. Not really sure what you mean, but this is my failed attempt. Maybe it's just an issue with the UI, but I don't see that cardinality can be omitted. Can you offer any insights? Thanks

![Image](https://github.com/user-attachments/assets/0dc12c60-2568-48aa-9304-2d349c1d75aa)

![Image](https://github.com/user-attachments/assets/d8410e13-9ccf-41f6-b330-6ec82fa21199)

![Image](https://github.com/user-attachments/assets/ee6fd9aa-0e6e-4be0-b5ef-24d760ee12ed)

![Image](https://github.com/user-attachments/assets/106d4121-41aa-462f-83b9-c853ac2d4e9f)

![Image](https://github.com/user-attachments/assets/02e87d6f-fbd5-4687-a97e-25a9d22bce80)

[IdsMaker_file.ids.txt](https://github.com/user-attachments/files/18725799/IdsMaker_file.ids.txt)

**Comment by Moult** [comment]
The IDS is invalid:

```
      <ids:requirements>
        <ids:entity cardinality="required">
...
```

As the error message (misleadingly, but correctly) says, the cardinality attribute is not in the schema. So remove `cardinality="required"` to make it valid.

You'll also need to remove the:

```
          <ids:predefinedType>
            <ids:simpleValue>USERDEFINED</ids:simpleValue>
          </ids:predefinedType>
```

If you want to check the predefined type via attributes.

**Comment by jwouellette** [comment]
So now I'm thoroughly confused.

1) What is the best method to indicate and check for a USERDEFINED value of PredefinedType? According to the "REQUIRED" text at the end of the "Entities" section https://github.com/buildingSMART/IDS/blob/development/Documentation/UserManual/facet-configurations.md, Rick's first (Spec 1.) should work and makes the most sense to an end user:

> [@atomczak](https://github.com/atomczak) [@Moult](https://github.com/Moult) I'm struggling with this issue right now. Can you explain the difference between these two IDS specifications?
> 
> Spec 1.
> ![Image](https://github.com/user-attachments/assets/3bc4b413-135a-4b06-871a-f2802f0cecb1)
> 
> Spec 2. 
> ![Image](https://github.com/user-attachments/assets/ec5f616c-87c4-4cfa-be96-e142f8ac08e8)
> 
> Given this entry in my IFC file
> 
> > #872=IFCBEAM('1L5c$BTMnCxhBJ7IXhvkC8',#14666,'Segment 1',$,'GFRP',#851,#871,$,.USERDEFINED.);
> 
> When checking my file against the IDS, the first specification signals an error, while the second specification indicates the requirement is satisfied.
> 
> What is the difference?

2) Why is the cardinality an error?

**Comment by RickBrice** [comment]
I think the cardinality is an error, because it's not in the IDS schema as part of the requirement. This comes from a bug in the bimVision IDS Maker tool.

![Image](https://github.com/user-attachments/assets/829b03e8-d659-4bc6-b5e4-6ea28097b3e0)

![Image](https://github.com/user-attachments/assets/98ca9b2b-5f70-4e56-8cff-6c16e9b9c7f9)

IDS Maker knows it is creating an invalid file, but there is no way to not specify a "Requirement type".

Open the same file in usBIM.IDSeditor, no error. No cardinality in the Requirements entity XML

![Image](https://github.com/user-attachments/assets/87f45db7-f19d-44d2-a168-5b84ead45a89)

**Comment by Moult** [comment]
The best way to check for userdefined is either by attribute facet or by enum restriction specifying known valid values.

Cardinality is invalid there because the XSD says so. There is a comment explaining it in the xsd but im on a phone right now.

Dion Moult

Sent from Proton Mail Android

-------- Original Message --------

**Comment by RickBrice** [comment]
I think what I'm learning here is that this not expressive enough.

![Image](https://github.com/user-attachments/assets/fb0ddb1b-b074-4a5b-89bf-38c6cf217112)

If the IfcBeam -> IfcRelDefinesByType -> IfcBeamType is used to define a beam, and IfcBeamType.PredefinedType = USERDEFINED then (1) IfcBeam.PredefinedType and IfcBeam.ObjectType can't be used and (2) the type is inherited from IfcBeamType.ElementType=GFRP

This method of specifying the custom type is more general and covers both cases.

![Image](https://github.com/user-attachments/assets/ca158e8f-718a-4bba-8cab-f0f965de7248)

**Comment by Moult** [comment]
Exactly. Inheritance and override is a core feature in IFC, not just for predefined types, but also for classifications, properties and materials. This is exactly the same in IDS. E.g. if you ask for a concrete object, a concrete occurence and an occurrence that inherits concrete from its type is identical.

**Comment by RickBrice** [comment]
https://github.com/buildingSMART/IDS/blob/dd7b0689a7ddd6e62133d0eaba2f662f743eee9c/Schema/ids.xsd#L120-L128

**Comment by vovka236** [comment]
> I think the cardinality is an error, because it's not in the IDS schema as part of the requirement. This comes from a bug in the bimVision IDS Maker tool.

Nobody likes bugs 🐞, neither do we!

This and a few other issues have already been located and fixed, the new version including also additional improvements is about to appear.
 
This afternoon, a new version of IDS Maker will be released 🚀

**Comment by atomczak** [comment]
This issue is solved with conlusion:
- predefined type of entity facet checks for user defined types too, no need to check if a value in IFC is set to USERDEFINED 
- don't use the predefined type of entity facet to check if the type is set to USERDEFINED; instead, use the attribute facet
- entity does not have cardinality (_IDS Maker_ bug that was fixed)
